In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [3]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv("lending_data.csv")

# Review the DataFrame
df.sample(10)

loan_size  interest_rate  borrower_income  debt_to_income  \
48254     9500.0          7.172            48100        0.376299   
53265     7400.0          6.289            39800        0.246231   
43881     6900.0          6.036            37400        0.197861   
68460     8900.0          6.892            45500        0.340659   
62869    11600.0          8.051            56400        0.468085   
20850     9000.0          6.966            46200        0.350649   
33484     9900.0          7.337            49600        0.395161   
12205    10300.0          7.514            51300        0.415205   
67503     8900.0          6.906            45600        0.342105   
75298    18100.0         10.834            82600        0.636804   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
48254                4                 0       18100            0  
53265                2                 0        9800            0  
43881                1                 0        7400            0  
68460                3                 0       15500            0  
62869                5                 1       26400            0  
20850                3                 0       16200            0  
33484                4                 0       19600            0  
12205                4                 1       21300            0  
67503                3                 0       15600            0  
75298               11                 2       52600            1

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
df.columns

Index(['loan_size', 'interest_rate', 'borrower_income', 'debt_to_income',
       'num_of_accounts', 'derogatory_marks', 'total_debt', 'loan_status'],
      dtype='object')

In [5]:
# Separate the data into labels and features
# Separate the y variable, the labels
y = df['loan_status']

# Separate the X variable, the features
X = df.drop(columns='loan_status')

In [6]:
# Review the y variable Series
y.sample(10)

45593    0
69788    0
6635     0
919      0
19020    0
51443    0
66320    0
76509    1
48349    0
51718    0
Name: loan_status, dtype: int64

In [7]:
# Review the X variable DataFrame
X.sample(10)

loan_size  interest_rate  borrower_income  debt_to_income  \
54718    11800.0          8.152            57300        0.476440   
62110    10900.0          7.768            53700        0.441341   
48686     9500.0          7.166            48000        0.375000   
32552     9600.0          7.193            48300        0.378882   
65268    10200.0          7.463            50800        0.409449   
36226     9100.0          7.006            46500        0.354839   
59855     8900.0          6.893            45500        0.340659   
36837     9500.0          7.180            48200        0.377593   
37091    10700.0          7.673            52800        0.431818   
20393     9300.0          7.083            47200        0.364407   

       num_of_accounts  derogatory_marks  total_debt  
54718                6                 1       27300  
62110                5                 1       23700  
48686                4                 0       18000  
32552                4                 0       18300  
65268                4                 1       20800  
36226                3                 0       16500  
59855                3                 0       15500  
36837                4                 0       18200  
37091                5                 1       22800  
20393                3                 0       17200

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [10]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [11]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [12]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(solver='lbfgs', max_iter=200, random_state=1)

# Fit the model using training data
model.fit(X_train, y_train)

LogisticRegression(max_iter=200, random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [34]:
# Make a prediction using the testing data
x_test_pred = model.predict(X_test)
df_results = pd.DataFrame({
    "test" : y_test,
    "pred" : x_test_pred
})
df_results.sample(10)

test  pred
25442     0     0
31939     0     0
75944     1     1
33779     0     0
66843     0     0
14737     0     0
54486     0     0
55350     0     0
66578     0     0
23330     0     0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [14]:
# Print the balanced_accuracy score of the model
from sklearn.metrics import accuracy_score
accuracy_score(y_test, x_test_pred)

0.9918489475856377

In [15]:
# Generate a confusion matrix for the model

confusion_matrix(y_test, x_test_pred)

array([[18663,   102],
       [   56,   563]])

In [18]:
# Print the classification report for the model

target_names = ['Healthy Loan', 'High-Risk Loan']
print(classification_report(y_test, x_test_pred, target_names = target_names))

                precision    recall  f1-score   support

  Healthy Loan       1.00      0.99      1.00     18765
High-Risk Loan       0.85      0.91      0.88       619

      accuracy                           0.99     19384
     macro avg       0.92      0.95      0.94     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Very well: The model has an accuracy of 99%, the precision of the "Healthy Loan" is 100% and the "High-Risk Loan" is 85%. Considering these metrics, the model predicts very well and with high precision.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points.

In [62]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
model_ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_train_resampled, y_train_resampled = model_ros.fit_resample(X_train, y_train)

In [68]:
# Count the distinct values of the resampled labels data
pd.Series(y_train_resampled).value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [69]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(solver='lbfgs', max_iter=200, random_state=1)

# Fit the model using the resampled training data
logistic_regression_model.fit(X_train_resampled, y_train_resampled)

# Make a prediction using the testing data
testing_prediction = logistic_regression_model.predict(X_test)
df_pred_results = pd.DataFrame({
    "test" : y_test,
    "pred" : testing_prediction
})
df_results.sample(10)

test  pred
37425     0     0
13389     0     0
24369     0     0
56297     0     0
29586     0     0
17285     0     0
75787     1     1
5118      0     0
40398     0     0
32398     0     0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [70]:
# Print the balanced_accuracy score of the model
from sklearn.metrics import accuracy_score
acc_score = accuracy_score(y_test, testing_prediction)
print(acc_score)

0.9938093272802311


In [71]:
# Generate a confusion matrix for the model

confusion_matrix(y_test, testing_prediction)

array([[18649,   116],
       [    4,   615]])

In [72]:
# Print the classification report for the model

target_names = ['Healthy Loan', 'High-Risk Loan']
testing_report = classification_report(y_test, testing_prediction, target_names = target_names)
print(testing_report)

                precision    recall  f1-score   support

  Healthy Loan       1.00      0.99      1.00     18765
High-Risk Loan       0.84      0.99      0.91       619

      accuracy                           0.99     19384
     macro avg       0.92      0.99      0.95     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Also very well: The model has an accuracy of 99%, the precision of the "Healthy Loan" is 100% and the "High-Risk Loan" is 84%. Considering these metrics, the model predicts very well and with high precision.